apps/omg_child_chain/lib/omg_child_chain/fees/file_adapter.ex

```diff
@@ -0,0 +1,75 @@
+# Copyright 2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.ChildChain.Fees.FileAdapter do
+  @moduledoc """
+  Adapter for fees stored in a JSON file (defined in omg_child_chain/priv config :omg_child_chain,
+  :fee_specs_file_name).
+  """
+  @behaviour OMG.ChildChain.Fees.Adapter
+
+  use OMG.Utils.LoggerExt
+
+  alias OMG.ChildChain.Fees.FeeParser
+
+  @doc """
+  Reads fee specification file if needed and returns its content.
+  When using this adapter, the operator can change the fees by updating a
+  JSON file that is loaded from disk (path variable).
+  """
+  # sobelow_skip ["Traversal"]
+  @impl true
+  def get_fee_specs(recorded_file_updated_at) do
+    path = get_path()
+
+    with {:changed, file_updated_at} <- check_file_changes(path, recorded_file_updated_at),
+         {:ok, content} <- File.read(path),
+         {:ok, fee_specs} <- FeeParser.parse(content) do
+      {:ok, fee_specs, file_updated_at}
+    else
+      {:unchanged, _last_changed_at} ->
+        :ok
+
+      error ->
+        error
+    end
+  end
+
+  defp check_file_changes(path, recorded_file_updated_at) do
+    actual_file_updated_at = get_file_last_modified_timestamp(path)
+
+    case actual_file_updated_at > recorded_file_updated_at do
+      true ->
+        {:changed, actual_file_updated_at}
+
+      false ->
+        {:unchanged, recorded_file_updated_at}
+    end
+  end
+
+  defp get_file_last_modified_timestamp(path) do
+    case File.stat(path, time: :posix) do
+      {:ok, %File.Stat{mtime: mtime}} ->
+        mtime
+
+      # possibly wrong path - returns current timestamp to force file reload where file errors are handled
+      _ ->
+        :os.system_time(:second)
+    end
+  end
+
+  defp get_path do
```

Comment:
Do you prefer brackets or no brackets?

Reply:
I prefer without if there are no arguments :) Can change if we follow a different style.

Reply:
Pls double check what it says.

Reply:
Can you be more explicit? I don't get it. Thought you meant parenthesis at first, now I'm not sure anymore. What's `it` also? Would be great to have less cryptic comments to avoid this kind of back and forth - if I did something wrong, just say so.

Reply:
hey! I had no bad intentions. My comment was only saying that you should look at the Elixir style guide we're using... I didn't know what the answer is as well! But now I checked: https://github.com/lexmag/elixir-style-guide#fun-parens

Reply:
@InoMurko Ah, I see, thanks! And, no worries, I was just super confused which led to some frustration. You're totally right about the style, I'll add the parenthesis to follow the guide :)

Reply:
I assumed for 2 full years that the style guide is to not have zero-arity fn parentheses :mindblown: